### DISCUSSION TITLE
How to Build a Generic Filter with Prisma in Node.js (with Nested Relations)?

### DISCUSSION THREAD

**Discussion by C-W-D-Harshit** [prisma/prisma#25107] [Q&A] [answered]
### Question

Hey everyone! 👋

I'm working on a **Node.js/TypeScript** project where I need to implement **generic filtering** for users with Prisma. The tricky part is that I need to apply filters dynamically, and some of the fields are nested relations, like filtering users based on projects or other related entities. I'm building an API to handle this, and the frontend sends filters in a flexible structure.

Here's an example of the type of filters I'm dealing with:

# Example Filter Request from Frontend:

    {
      "filter": {
        "name": "John",
        "role": "ADMIN",
        "projects": {
          "name": "Project A"
        }
      }
    }

I want to build a backend that can handle these filters dynamically using Prisma, where the fields and the conditions can change based on the request. I've already made some progress by writing a function to construct a `where` clause dynamically.

# My Current Backend Code:
```js
    import { PrismaClient } from '@prisma/client';
    
    const prisma = new PrismaClient();
    
    const buildWhereClause = (filters: any) => {
      const whereClause: any = {};
    
      for (const [key, value] of Object.entries(filters)) {
        if (typeof value === 'object' && !Array.isArray(value)) {
          whereClause[key] = {
            some: buildWhereClause(value) // Nested relations like 'projects'
          };
        } else {
          whereClause[key] = value; // Direct field filters like 'name' or 'role'
        }
      }
    
      return whereClause;
    };
    
    const getUsersWithFilters = async (filter: any) => {
      const whereClause = buildWhereClause(filter);
    
      return await prisma.user.findMany({
        where: whereClause,
        include: { projects: true }
      });
    };
```
# How the API Works:

* The frontend sends a filter object in JSON.
* The backend dynamically builds the `where` clause using a recursive function for handling nested relations.
* Prisma executes the query, including any related entities (e.g., projects).

# What I’m Looking for:

I want to make this filtering more **generic and flexible**:

1. **Handling multiple conditions like** `AND`\*\*,\*\* `OR`\*\*,\*\* `IN`\*\*,\*\* `LIKE`\*\*, etc.\*\* For example, I want to support cases where the frontend sends complex conditions like:{ "filter": { "AND": \[ { "role": "ADMIN" }, { "projects": { "name": "Project A" } } \], "name": { "contains": "John" } } }

**Better handling of nested relations**, like `projects`, `tasks`, etc. — where I might want to filter based on multiple nested conditions.

* **Handling different types of comparisons** (`contains`, `startsWith`, `gte`, `lte`, etc.) dynamically based on the request.

# Example of a Complex Request:

    {
      "filter": {
        "AND": [
          { "role": "ADMIN" },
          { "projects": { "name": "Project A" } }
        ],
        "name": { "contains": "John" },
        "age": { "gte": 30 },
        "email": { "contains": "example.com" }
      }
    }

Has anyone done something similar with Prisma? I'd love to see examples or hear suggestions on how I can improve this. Any advice on **building dynamic filtering systems** would be really appreciated!

Thanks in advance!

**Comment by galenwinsor** [comment]
I think this is totally doable.

First of all, I wouldn't type the where clause as `any` and would do some validation to make sure the fields in the filter are valid. Your frontend logic probably makes sure the JSON is in the right shape, but adding some typing on the backend will give you some more peace of mind in that it'll force you to do validation. I personally use `zod` for this kind of thing. I can expand on this if you want.

Beyond that, I don't see how your current solution doesn't already support the use cases you're describing. Your builder function should handle that second example filter object you're showing, since it just maps keys to nested objects. **The core issue here is validating that you're sending a properly formatted JSON,** and typing things will force you to do that.

Am I understanding your question? Or do you mean something else by "generic?"

**Reply by C-W-D-Harshit** [reply]
You are correct; my current code does work, but I was unsure if it was the best approach. As you mentioned, typing the whereClause would indeed provide peace of mind and ensure that the fields in the filter are valid. I agree that adding types will enforce better validation and help in managing properly formatted JSON.

When I referred to "generic," I meant that I want the builder function to be adaptable and reusable in different contexts. I understand that typing the whereClause may be more complex, but it seems like a worthwhile investment for robust validation and type safety.

Thank you for pointing out the importance of typing and validation; I will consider using `zod` or similar tools to improve this aspect of my code. Could you provide a simple, working code example to demonstrate how to implement this approach effectively?

**Reply by galenwinsor** [reply, marked as answer]
Hi @C-W-D-Harshit, sorry for the late reply. Sure thing.

It could get a little complicated, but the basic setup is this:

```typescript
const WhereClauseSchema = z.object({
  where: z.object({
    uid: z.string(),
    // ... you define any other optional fields you might want to validate
  })
// `satisfies` ensures that the schema matches the type of the query
}) satisfies z.Schema<Prisma.WhateverTableFindUniqueArgs>;

const result = WhereClauseSchema.safeParse(jsonObject);

if (!result.success) {
  console.error(result.error);
} else {
  // do something with result.data
}
```

Prisma has [some documentation](https://www.prisma.io/docs/orm/prisma-client/queries/custom-validation) on custom validation.

**Reply by galenwinsor** [reply]
In this case, `result.data` will be typed as having the shape `{ where: { uid: string } }`, even if the json object includes more fields at runtime. Zod will give you type safety for the fields you specify in the schema.

**Reply by C-W-D-Harshit** [reply]
Thanks, @galenwinsor, for your answer! This is some cool stuff!
